packages/actor-optimize-link-traversal-filter-tree-links/lib/ActorOptimizeLinkTraversalFilterTreeLinks.ts

```diff
@@ -0,0 +1,203 @@
+      const relevantQuads = this.findRelavantQuad(queryBody, relation.path.value);
+      if (relevantQuads.length === 0) {
+        filterMap.set(relation.node, true);
+        continue;
+      }
+
+      // Create the binding in relation to the relevant quad
+      const bindings = this.createBinding(relevantQuads, relation.value.quad);
+      const filterExpression: Algebra.Operation = this.deleteUnrelevantFilter(filterOperation, bindings);
+      if (filterExpression.args.length === 0) {
+        filterMap.set(relation.node, true);
+        continue;
+      }
+      const evaluator = new AsyncEvaluator(filterExpression);
+      // Evaluate the filter with the relevant quad binding
+      const result: boolean = await evaluator.evaluateAsEBV(bindings);
+      filterMap.set(relation.node, result);
+    }
+    return { filters: filterMap };
+  }
+
+  /**
+   *
+   * @param queryBody
+   * @param path
+   * @returns RDF.Quad[]
+   * find the quad that has as predicate a the TREE:path of a relation
+   */
+  private findRelavantQuad(queryBody: RDF.Quad[], path: string): RDF.Quad[] {
+    const resp: RDF.Quad[] = [];
+    for (const quad of queryBody) {
+      if (quad.predicate.value === path && quad.object.termType === 'Variable') {
+        resp.push(quad);
+      }
+    }
+    return resp;
+  }
+
+  /**
+   * @param filterExpression
+   * @param binding
+   * @returns Algebra.Expression
+   * delete the filters that are not related to TREE relation
+   */
+  private deleteUnrelevantFilter(filterExpression: Algebra.Expression, binding: Bindings): Algebra.Expression {
+    let newFilterExpression: Algebra.Expression = {
+      expressionType: Algebra.expressionTypes.OPERATOR,
+      operator: filterExpression.operator,
+      type: Algebra.types.EXPRESSION,
+      args: [],
+    };
+    if ('operator' in filterExpression.args[0]) {
+      newFilterExpression.args = (<Algebra.Expression[]>filterExpression.args).filter(expression => {
+        for (const arg of expression.args) {
+          if ('term' in arg && arg.term.termType === 'Variable') {
+            return binding.has(arg.term.value);
+          }
+        }
+        return false;
+      });
+      if (newFilterExpression.args.length === 1) {
+        newFilterExpression = newFilterExpression.args[0];
+      }
+    } else {
+      for (const arg of (<Algebra.Expression[]>filterExpression.args)) {
+        if ('term' in arg && arg.term.termType === 'Variable' && !binding.has(arg.term.value)) {
+          newFilterExpression.args = [];
+          break;
+        }
+        newFilterExpression.args.push(arg);
+      }
+    }
+    return newFilterExpression;
+  }
+
+  /**
+   *
+   * @param relevantQuad
+   * @param relationValue
+   * @returns Bindings
+   * create the binding from quad related to the TREE:path
+   */
+  private createBinding(relevantQuad: RDF.Quad[], relationValue: RDF.Quad): Bindings {
+    let binding: Bindings = new BindingsFactory().bindings();
```

Comment:
I typically create once instance of this factory outside of the actor class (at the top of the file).
`const BF = new BindingsFactory();`